modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/ItIndexNodeTest.java

```diff
@@ -0,0 +1,318 @@
+    private static final int PART_ID = 0;
+
+    private String nodeName;
+
+    private IgniteServer node;
+
+    private Ignite ignite;
+    private InternalTable internalTable;
+    private int hashIdx;
+    private int sortIdx;
+
+    TableViewInternal tableViewInternal;
+
+    @BeforeEach
+    void setUp(TestInfo testInfo) {
+        nodeName = testNodeName(testInfo, 0);
+
+        node = TestIgnitionManager.start(nodeName, NODE_CONFIG, workDir.resolve(nodeName));
+
+        InitParameters parameters = new InitParametersBuilder()
+                .clusterName("cluster")
+                .metaStorageNodes(node)
+                .build();
+
+        node.initCluster(parameters);
+
+        assertThat(node.waitForInitAsync(), willCompleteSuccessfully());
+
+        ignite = node.api();
+
+        // Create zone with 1 partition.
+        String createZoneSql = "CREATE ZONE IF NOT EXISTS test_zone (PARTITIONS 1, REPLICAS 1) STORAGE PROFILES ['default']";
+
+        String createTableSql = "CREATE TABLE " + TABLE_NAME + "(\n"
+                + "    key int PRIMARY KEY,\n"
+                + "    field1   int,\n"
+                + "    field2   int\n"
+                + ") ZONE test_zone";
+        String createHashIndexSql = "CREATE INDEX IF NOT EXISTS " + HASH_IDX + " ON " + TABLE_NAME + " USING HASH (field1)";
+        String createSortIndexSql = "CREATE INDEX IF NOT EXISTS " + SORT_IDX + " ON " + TABLE_NAME + " USING SORTED (field2)";
+        String insertWithParams = "INSERT INTO " + TABLE_NAME + " (key, field1, field2) VALUES (?, ?, ?)";
+
+        ignite.sql().execute(null, createZoneSql);
+        ignite.sql().execute(null, createTableSql);
+        ignite.sql().execute(null, createHashIndexSql);
+        ignite.sql().execute(null, createSortIndexSql);
+        ignite.sql().execute(null, insertWithParams, 1, VALUE, VALUE);
+
+        IgniteImpl igniteImpl = unwrapIgniteImpl(ignite);
+
+        CatalogManager catalogManager = igniteImpl.catalogManager();
+
+        Catalog catalog = catalogManager.activeCatalog(igniteImpl.clock().nowLong());
+
+        hashIdx = getIndexIdBy(HASH_IDX, catalog);
+        sortIdx = getIndexIdBy(SORT_IDX, catalog);
+
+        // Get the public Table API.
+        Table table = ignite.tables().table(TABLE_NAME);
+
+        assertNotNull(table);
+
+        // Unwrap to get TableViewInternal.
+        tableViewInternal = unwrapTableViewInternal(table);
+
+        // Get the InternalTable.
+        internalTable = tableViewInternal.internalTable();
+
+        assertNotNull(internalTable);
+    }
+
+    @AfterEach
+    void tearDown() {
+        node.shutdown();
+    }
+
+    @Test
+    void testHashIndexThrowsWithRangeScan() {
+        Publisher<BinaryRow> publisher =  internalTable.scan(PART_ID, null, hashIdx,
+                createMatchingRange());
+        CompletableFuture<List<BinaryRow>> resultFuture = subscribeToList(publisher);
+        assertThrows(ExecutionException.class, () -> {
```

Comment:
Here and below, we should not only check that an exception is thrown, but also verify that it's an exception of proper class with proper error message. You may find the `org.apache.ignite.internal.testframework.IgniteTestUtils#assertThrowsWithCause(org.apache.ignite.internal.lang.RunnableX, java.lang.Class<? extends java.lang.Throwable>, java.lang.String)` helpful. Please pay attention to the last String param, that is a prefix of an expected message.

Here's an example
```
        assertThrowsWithCause(() -> marshaller.marshal(tuple1), InvalidTypeException.class,
                "Value too long [column='BYTES', type=BYTE_ARRAY(5)]"
        );
```